scripts/gas-refund-program/transactions-indexing/successful-swap-tx-indexing.ts

```diff
@@ -0,0 +1,170 @@
+import { getSwapsForAccounts } from './swaps-subgraph';
+import {
+  getRefundPercent,
+  PendingEpochGasRefundData,
+} from '../../../src/lib/gas-refund';
+
+const PARTITION_SIZE = 1000; // depends on thegraph capacity and memory
+
+export async function computeSuccessfulSwapsTxFeesRefund({
+  chainId,
+  startTimestamp,
+  endTimestamp,
+  pspNativeCurrencyDailyRate,
+  epoch,
+  stakes,
+}: {
+  chainId: number;
+  startTimestamp: number;
+  endTimestamp: number;
+  pspNativeCurrencyDailyRate: HistoricalPrice;
+  epoch: number;
+  stakes: StakedPSPByAddress;
+}): Promise<void> {
+  const logger = global.LOGGER(
+    `GRP:TRANSACTION_FEES_INDEXING: epoch=${epoch}, chainId=${chainId}`,
+  );
+
+  const blockInfo = BlockInfo.getInstance(chainId);
+  const [epochStartBlock, epochEndBlock] = await Promise.all([
+    blockInfo.getBlockAfterTimeStamp(startTimestamp),
+    blockInfo.getBlockAfterTimeStamp(endTimestamp),
+  ]);
+  const findSameDayPrice = constructSameDayPrice(pspNativeCurrencyDailyRate);
+  const stakersAddress = Object.keys(stakes);
+
+  assert(
+    epochStartBlock,
+    `no start block found for chain ${chainId} for timestamp ${startTimestamp}`,
+  );
+  assert(
+    epochEndBlock,
+    `no start block found for chain ${chainId} for timestamp ${endTimestamp}`,
+  );
+
+  logger.info(
+    `swapTracker start indexing between ${epochStartBlock} and ${epochEndBlock}`,
+  );
+
+  let [accumulatedTxFeesByAddress, veryLastBlockNumProcessed] =
+    await readPendingEpochData({
+      chainId,
+      epoch,
+    });
+
+  const startBlock = Math.max(epochStartBlock, veryLastBlockNumProcessed + 1);
+
+  logger.info(`start processing at block ${startBlock}`);
+
+  for (
+    let _startBlock = startBlock;
+    _startBlock < epochEndBlock;
+    _startBlock += PARTITION_SIZE
+  ) {
+    const _endBlock = Math.min(_startBlock + PARTITION_SIZE, epochEndBlock);
+
+    logger.info(
+      `start indexing partition between ${_startBlock} and ${_endBlock}`,
+    );
+
+    const swaps = await getSwapsForAccounts({
+      startBlock: _startBlock,
+      endBlock: _endBlock,
+      accounts: stakersAddress,
+      chainId,
+    });
+
+    logger.info(
+      `fetched ${swaps.length} swaps withing startBlock ${_startBlock} and endBlock ${_endBlock}`,
+    );
+
+    accumulatedTxFeesByAddress = swaps.reduce<TxFeesByAddress>((acc, swap) => {
+      const address = swap.txOrigin;
+
+      const swapperAcc = acc[address];
+
+      const pspRateSameDay = findSameDayPrice(swap.timestamp);
+
+      if (!pspRateSameDay) {
+        logger.warn(
+          `Fail to find price for same day ${
+            swap.timestamp
+          } and rates=${JSON.stringify(
+            pspNativeCurrencyDailyRate.flatMap(p => p.timestamp),
+          )}`,
+        );
+
+        return acc;
+      }
+
+      const currGasUsed = new BigNumber(swap.txGasUsed.toString());
+      const accGasUsed = currGasUsed.plus(swapperAcc?.accumulatedGasUsed || 0);
```

Comment:
optional chaining is not needed here and in other places with `swapperAcc`

Reply:
It is needed. The type hint you get is wrong. swapperAcc is undefined on visiting an address first time.